snooty/semanticparser.py

```diff
@@ -54,3 +55,87 @@ def slug_title(self, pages: Dict[FileId, Page]) -> Dict[str, SerializableType]:
             slug_title_dict[slug] = title
 
         return {"slugToTitle": slug_title_dict}
+
+    def toctree(self, pages: Dict[FileId, Page]) -> Dict[str, SerializableType]:
+        fileid_dict = {}
+        toctree: Dict[str, Any] = {"toctree": {}}
+
+        starting_fileid: FileId = [
+            fileid
+            for fileid in pages.keys()
+            if str(fileid) == "contents.txt" or str(fileid) == "index.txt"
+        ][0]
+
+        # Construct a {slug: fileid} mapping so that we can retrieve the full file name
+        # given a slug. We cannot use the with_suffix method since the type of the slug
+        # in find_toctree_nodes(...) is string rather than FileId.
+        for fileid in pages:
+            slug = fileid.without_known_suffix
+            fileid_dict[slug] = fileid
+
+        slug_title: Dict[str, Any] = cast(Dict[str, Any], self.slug_title(pages))[
+            "slugToTitle"
+        ]
+
+        # Build the toctree
+        root: Dict["str", Any] = {}
+        ast: Dict[str, Any] = cast(Dict[str, Any], pages[starting_fileid].ast)
+        find_toctree_nodes(starting_fileid, ast, pages, root, fileid_dict, slug_title)
+
+        toctree["toctree"] = root
+        toctree["toctree"]["title"] = self.project_config.name
+        toctree["toctree"]["slug"] = "/"
+
+        return toctree
+
+
+# find_toctree_nodes is a helper function for SemanticParser.toctree that recursively builds the toctree
+def find_toctree_nodes(
+    fileid: FileId,
+    ast: Dict[str, Any],
+    pages: Dict[FileId, Page],
+    node: Dict[Any, Any],
+    fileid_dict: Dict[str, FileId],
+    slug_title: Dict[str, Any],
+) -> None:
+
+    # Base case: create node in toctree
+    if "children" not in ast.keys():
+        if node:
+            node["slug"] = fileid.without_known_suffix
+            node["title"] = slug_title[node["slug"]]
+        return
+
+    if ast["type"] == "directive":
+        if ast["name"] == "toctree" and "entries" in ast.keys():
+            if "children" not in node:
+                node["children"] = ast["entries"]
+            else:
+                node["children"].extend(ast["entries"])
+
+            # Recursively build the tree for each toctree node in this entries list
+            for toctree_node in node["children"]:
+                if "slug" in toctree_node:
+                    # Only recursively build the tree for internal links
+                    slug = toctree_node["slug"]
+                    if slug[0] == "/":
+                        slug = slug[1:]
+                    idx = slug.find(".")
```

Comment:
I believe what you're trying to do here is identify a link as a URL if it contains a `.`. That is a bit risky because filenames can actually include periods, which they often do when dealing with MongoDB numbered versions ([example](https://github.com/mongodb/docs/blob/master/source/includes/extracts-4.2-changes.yaml)). I had to approach this problem too, and ended up looking for the angle brackets that must enclose a URL to identify it—[see here](https://github.com/mongodb/snooty-parser/blob/master/snooty/rstparser.py#L582-L585).

Let me know if I've misunderstood what's going on here! And some comments when you update would be appreciated ☺️

Reply:
I'm actually removing the slug's file extension here! I'm doing it this way instead of using Raymund's `without_known_suffix` function since by this point in the code, I don't have the appropriate `FileId` object, just the slug as a string.

Slugs and urls are added to the toctree in lines 111-112:
`if "children" not in node:`
     `node["children"] = ast["entries"]`

I'll add some comments to make this clear!

Reply:
Just realized this doesn't work because of filenames like `security-x.509.txt` 😬 I'll push a modification soon!

Reply:
Oh interesting! Is this something you ran into a lot? I think the vast majority of the time, slugs are enumerated without extensions.

In any case, this is something we will want to handle in the regular parser, not the semantic layer. Just include the slugs as-is here!

Reply:
Discussed offline: we will continue to strip file extensions from slugs in the semantic layer so that we can successfully find a slug's FileId object (currently we use a `{slug: FileId}` mapping to do so). This logic will be moved to the parse layer in [DOCSP-7595](https://jira.mongodb.org/browse/DOCSP-7595).